### PR TITLE
Use vtkmodules

### PR DESCRIPTION
**Pull request by akaszynski**
### Use `vtkmodules`

This PR proposes that we use `vtkmodules` instead of importing everything from vtk.  I didn't want to break backwards compatibility with VTK 8.x and earlier, so I implemented this such that all the commonly used imports are in a new module named `_vtk.py`.  This accomplishes a few things:

- Deals with the annoying lint errors complaining about how `vtk` doesn't contain attribute x.
- Allows us to selectively load only the libraries that are needed within `pyvista`.  The vast majority (98%) are loaded on init since the load times aren't too bad, but a few are imported lazily (like the rarely used `vtkGL2PSExporter`).  We can always improve this down the line, but for now I think this is a good start.
- Improves load time and memory usage

#### pyvista/master

```
time python -c "import pyvista"

real	0m0.327s
user	0m0.564s
sys	0m0.410s
```

#### This branch:

```
time python -c "import pyvista"

real	0m0.283s
user	0m0.542s
sys	0m0.385s
```

Load time variation is even better on Windows:

```py
timeit os.system('python -c "import pyvista"')
```

#### master
```
595 ms ± 1.43 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

#### this branch

```
485 ms ± 1.72 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

---
### Memory Usage
Memory usage also shows some significant improvements:

```
valgrind python -c "import pyvista"
```

#### master
```
 total heap usage: 47,253 allocs, 41,158 frees, 79,869,570 bytes allocated
```

#### this branch
```
total heap usage: 27,436 allocs, 23,009 frees, 50,234,867 bytes allocated
```

### CODE REVIEW

**Comment by akaszynski** [comment]
@pyvista/developers, this is ready for review.

There's a substantial change to the code base here, and I'd like to know your opinions about this PR.

**Review by adeak — COMMENTED** [review]
Is it a conscious choice that proper `vtk` imports are kept in examples, doctests and tests? I guess the first two of the three are straightforward, because they are user-facing. Tests could go either way perhaps.

File: pyvista/plotting/camera.py
Comment (akaszynski) [reply pyvista/plotting/camera.py:5]: Oops!  Thanks for finding that.

**Comment by akaszynski** [comment]
> Is it a conscious choice that proper `vtk` imports are kept in examples, doctests and tests? I guess the first two of the three are straightforward, because they are user-facing. Tests could go either way perhaps.

I was originally going to go with `vtki` (like vtk imports), but went with `_vtk` as this is something I want hidden from the user so as not to confuse them.  What we're doing here is specific to `pyvista` since we're only exposing/extending certain functionality from `vtk`.  I don't want to take away from the original module.

True, tests can go either way, but I'd like to keep with the `vtk` module since we're outside the `pyvista` module and `_vtk` is now private (well, as private as you can make it in Python).



> And there seems to be a cached test image that changed a bit: `test_add_point_labels()` in `tests/plotting/test_plotting.py`. I think the zoom (or label sizes) changed a bit. The error is in the "warning" region.

That needs to be fixed in a separate PR.  The images are very similar, and I think the variation is due to off vs. on screen or version variations.  I might permit certain tests to have larger allowable errors or just ignore this test.

**Comment by adeak** [comment]
Makes sense (all three things, in fact). Thanks.

File: .gitignore
Comment (adeak) [inline .gitignore:8]: What generates these?

File: pyvista/_vtk.py
Comment (adeak) [inline pyvista/_vtk.py:339]:
```suggestion
```
Missed this duplication on my first skim, sorry.

The style for the imports (empty lines, placement of closing parentheses)  might have to be unified before the PR is merged, but this is just bikeshedding. Heroic work you've done on this!

**Comment by adeak** [comment]
Breaking test is the same as https://github.com/pyvista/pyvista/pull/1150#issuecomment-776968117.

File: .gitignore
Comment (akaszynski) [reply .gitignore:8]: Emacs generates `#*#` files either as back files when editing or when not saving changes when exiting.  I should have annotated that in the `.gitignore`

File: pyvista/_vtk.py
Comment (akaszynski) [reply pyvista/_vtk.py]: Good catch.

> The style for the imports (empty lines, placement of closing parentheses) might have to be unified before the PR is merged

I'd like to keep the imports on this module one per line for readability since we're placing a ton.  As for empty lines, there might have been a method to the madness at some point; does it make sense to remove all empty lines?  We can do that and since I'm waiting for @banesullivan to review, we should clean this up.

Comment (adeak) [reply pyvista/_vtk.py]: One on each line is good, I agree. There are two violations of this rule:
```sh
$ git grep -nE '^\s*from.+,[^$]' pyvista/_vtk.py 
pyvista/_vtk.py:44:    from vtkmodules.vtkIOPLY import vtkPLYReader, vtkPLYWriter
pyvista/_vtk.py:221:    from vtkmodules.vtkCommonMath import vtkMatrix4x4, vtkMatrix3x3
```

I think both "always empty lines" and "no empty lines" would work.

And similarly, I don't have a very strong opinion about closing parentheses, but now there are three styles being mixed:
```py
    from vtkmodules.vtkRenderingOpenGL2 import (vtkOpenGLHardwareSelector,
                                                vtkRenderStepsPass,
                                                vtkEDLShading,
                                                vtkOpenGLRenderer,
                                                vtkOpenGLTexture)
    [snip]
    from vtkmodules.vtkIOImage import (vtkBMPReader,
                                       vtkDEMReader,
                                       vtkDICOMImageReader,
                                       vtkDICOMImageReader,
                                       vtkJPEGReader,
                                       vtkJPEGReader,
                                       vtkMetaImageReader,
                                       vtkNrrdReader,
                                       vtkNrrdReader,
                                       vtkPNGReader,
                                       vtkPNMReader,
                                       vtkSLCReader,
                                       vtkTIFFReader,
                                       vtkTIFFReader,
    )
    [snip]
    from vtkmodules.vtkRenderingAnnotation import (vtkScalarBarActor,
                                                   vtkCornerAnnotation,
                                                   vtkAxesActor,
                                                   vtkAnnotatedCubeActor,
                                                   vtkLegendBoxActor,
                                                   vtkCubeAxesActor,
                                                   )

Comment (adeak) [reply pyvista/_vtk.py]: If I were being pedantic I think this is the part of PEP 8 that is relevant (excerpt):
> The closing brace/bracket/parenthesis on multiline constructs may either line up under the first non-whitespace character of the last line of list, as in:
> 
> ```
> my_list = [
>     1, 2, 3,
>     4, 5, 6,
>     ]
> ```
> or it may be lined up under the first character of the line that starts the multiline construct, as in:
> ```
> my_list = [
>     1, 2, 3,
>     4, 5, 6,
> ]
> ```

But arguably imports would look weirder, being pulled in to the left like that. So I think the third option above is fine.

**Comment by akaszynski** [comment]
> Breaking test is the same as [#1150 (comment)](https://github.com/pyvista/pyvista/pull/1150#issuecomment-776968117).

No longer checking this in the doctest.  I usually don't end up with any issues when using the `clean` filter, so it may be isolated to this specific example.
https://github.com/pyvista/pyvista/blob/ba6055825da1e898fb301107f545801cc28eb24c/pyvista/core/filters.py#L3814

File: .gitignore
Comment (adeak) [inline .gitignore:64]: [citation needed]

Comment (akaszynski) [reply .gitignore:64]: Haha!  Glad someone noticed that.  Figured I keep that in until someone noticed.  It took 15 minutes.

Comment (adeak) [reply .gitignore]: It was a nice try ;)

**Comment by akaszynski** [comment]
Pinging @pyvista/developers.  This is ready for review and I'd like to have this in `master` so we can have this merged for a while before the release.

**Review by tkoyama010 — APPROVED** [review]
LGTM after resolving the conflict.

File: pyvista/__init__.py
Comment (larsoner) [inline pyvista/__init__.py:25]: Move to the top with the rest of the `from pyvista` lines?

File: pyvista/_vtk.py
Comment (larsoner) [inline pyvista/_vtk.py:329]: Down the road I could see this pattern confusing some future dev who tries `if isinstance(x, vtkGL2PSExporter)` for example (or anything else that assumes `vtkGL2PSExporter` is the `vtkmodules.vtkIOExportGL2PS.vtkGL2PSExporter` class) and it doesn't do what it seems like it should. I would name these something more like `_lazy_vtkGL2PSExporter` or something to make it clear it's not actually the vtk class itself.

File: pyvista/_vtk.py
Comment (larsoner) [inline pyvista/_vtk.py:319]: This appears to be identical to the import for VTK < 9 -- and same goes at least for the `numpy_support` imports. It would be nice to unindent the imports that are common to both code paths to make it clear what is version dependent and what's not

**Review by larsoner — COMMENTED** [review]
Other than a couple of minor comments and needing to fix a merge conflict, LGTM

File: pyvista/_vtk.py
Comment (akaszynski) [reply pyvista/_vtk.py:319]: The problem with using `vtk` instead of `vtkmodules` is importing anything from `vtk` loads the top level module, which in the case of `vtk`, means loading all the vtk libraries.  Do demonstrate, I first ran `time python -c "import pyvista"`:
```
time python -c "import pyvista"

real	0m0.289s
user	0m0.494s
sys	0m0.357s
```
Next, I placed the following at the top of the module:

```py
from vtk.numpy_interface.dataset_adapter import (VTKObjectWrapper,
                                                 VTKArray,
                                                 numpyTovtkDataArray)
```

which results in

```
time python -c "import pyvista"

real	0m0.334s
user	0m0.510s
sys	0m0.467s
```

If you go through `importime`, you'll also see all the modules being loaded.  Sadly, this won't work as a solution.

---

It's annoying for sure, but I think it's a far better solution rather than dropping vtk 8 support entirely, especially given the problems with closing render windows on vtk 9 on Linux.

File: pyvista/_vtk.py
Comment (akaszynski) [reply pyvista/_vtk.py:329]: Agreed, and I felt a little "lazy" about the way I implemented this.  I'll patch this.

Comment (larsoner) [reply pyvista/_vtk.py:329]: Oh I totally missed that this is `from vtkmodules` and below it's `from vtk`!

File: pyvista/__init__.py
Comment (akaszynski) [reply pyvista/__init__.py]: Yes, this needs to be moved up.  I'll resolve this.

File: pyvista/_vtk.py
Comment (akaszynski) [reply pyvista/_vtk.py:329]: I think this comment was meant for the other thread, right?

Comment (larsoner) [reply pyvista/_vtk.py:329]: Whoops, yep!

**Review by adeak — APPROVED** [review]

File: tests/test_init.py
Comment (akaszynski) [reply tests/test_init.py:15]: Yep, it's missing.  Thanks for catching that.

File: pyvista/plotting/camera.py
Comment (akaszynski) [reply pyvista/plotting/camera.py]: You've enlightened me.  I've had almost no knowledge of `__all__`, and learned (as I do with everything else) about it on SO here: https://stackoverflow.com/a/35710527/3369879

Regardless, I fixed this in a310e3849105813eeec7441c86f7d57b962da1f2

File: pyvista/core/filters.py
Comment (akaszynski) [reply pyvista/core/filters.py]: Fixed in a310e3849105813eeec7441c86f7d57b962da1f2

File: pyvista/plotting/plotting.py
Comment (akaszynski) [reply pyvista/plotting/plotting.py:26]: This must have been missed in our unit testing.  See:
https://github.com/pyvista/pyvista/blob/1720900d2849d9a9b80cf8b519b78c953fc45851/pyvista/plotting/plotting.py#L1623

Comment (adeak) [reply pyvista/plotting/plotting.py:26]: Yeah, I saw that it was used, that's why I got curious. I guess I could've just checked the coverage files I already have lying around:
```
>             if isinstance(texture, np.ndarray):
!                 texture = numpy_to_texture(texture)
>             if not isinstance(texture, (vtk.vtkTexture, vtk.vtkOpenGLTexture)):
!                 raise TypeError(f'Invalid texture type ({type(texture)})')
>             if mesh.GetPointData().GetTCoords() is None:
!                 raise ValueError('Input mesh does not have texture coordinates to support the texture.')
```

File: pyvista/_vtk.py
Comment (akaszynski) [reply pyvista/_vtk.py]: Fixed in fb3f18d.  It's not perfect, but these lazy loaders are only used for a handful of readers.

**Pull request merged**
